### DISCUSSION TITLE
Detect Mouse and keyboard events globally

### DISCUSSION THREAD

**Discussion by alihassan143** [halildurmus/win32#923] [Q&A]
@halildurmus is there is way to detect mouse movement and mouse clicks and any keyboard button click events

**Comment by alihassan143** [comment]
```dart
static bool _isMonitoring = false;
  static int _keyboardHook = 0;
  static int _mouseHook = 0;

  //

  
  
 static void startMonitoring() {
    if (_isMonitoring) return;
    _isMonitoring = true;

    // Set keyboard hook
    _keyboardHook = SetWindowsHookEx(
      WINDOWS_HOOK_ID.WH_KEYBOARD_LL,
      Pointer.fromFunction<HOOKPROC>(_keyboardProc, 0),
      GetModuleHandle(nullptr),
      0,
    );

    if (_keyboardHook == 0) {
      debugPrint("Failed to set keyboard hook: ${GetLastError()}");
      _isMonitoring = false;
      return;
    }

    // Set mouse hook
    _mouseHook = SetWindowsHookEx(
      WINDOWS_HOOK_ID.WH_MOUSE_LL,
      Pointer.fromFunction<HOOKPROC>(_mouseProc, 0),
      GetModuleHandle(nullptr),
      0,
    );

    if (_mouseHook == 0) {
      UnhookWindowsHookEx(_keyboardHook); // Clean up if mouse hook fails
      debugPrint("Failed to set mouse hook: ${GetLastError()}");
      _isMonitoring = false;
      return;
    }

    debugPrint("Monitoring started.");
  }

  static void stopMonitoring() {
    if (!_isMonitoring) return;
    _isMonitoring = false;

    if (_keyboardHook != 0) {
      UnhookWindowsHookEx(_keyboardHook);
      _keyboardHook = 0; // Reset the hook
    }

    if (_mouseHook != 0) {
      UnhookWindowsHookEx(_mouseHook);
      _mouseHook = 0; // Reset the hook
    }

    debugPrint("Monitoring stopped.");
  }

  static int _keyboardProc(int nCode, int wParam, int lParam) {
    debugPrint("Keyboard event: wParam=$wParam, lParam=$lParam");
    if (nCode >= 0) {
      debugPrint("Keyboard event: wParam=$wParam, lParam=$lParam");
      _keyBoardState.value = KeyboardState.active;
    }
    return CallNextHookEx(_keyboardHook, nCode, wParam, lParam);
  }

  static int _mouseProc(int nCode, int wParam, int lParam) {
    if (nCode >= 0) {
      debugPrint("Mouse event: wParam=$wParam, lParam=$lParam");
      _mouseState.value = MouseState.active;
    }
    return CallNextHookEx(_mouseHook, nCode, wParam, lParam);
  }
```


I am not getting any event what will be the issue ?

**Reply by alihassan143** [reply]
> ```dart
> static bool _isMonitoring = false;
>   static int _keyboardHook = 0;
>   static int _mouseHook = 0;
> 
>   //
> 
>   
>   
>  static void startMonitoring() {
>     if (_isMonitoring) return;
>     _isMonitoring = true;
> 
>     // Set keyboard hook
>     _keyboardHook = SetWindowsHookEx(
>       WINDOWS_HOOK_ID.WH_KEYBOARD_LL,
>       Pointer.fromFunction<HOOKPROC>(_keyboardProc, 0),
>       GetModuleHandle(nullptr),
>       0,
>     );
> 
>     if (_keyboardHook == 0) {
>       debugPrint("Failed to set keyboard hook: ${GetLastError()}");
>       _isMonitoring = false;
>       return;
>     }
> 
>     // Set mouse hook
>     _mouseHook = SetWindowsHookEx(
>       WINDOWS_HOOK_ID.WH_MOUSE_LL,
>       Pointer.fromFunction<HOOKPROC>(_mouseProc, 0),
>       GetModuleHandle(nullptr),
>       0,
>     );
> 
>     if (_mouseHook == 0) {
>       UnhookWindowsHookEx(_keyboardHook); // Clean up if mouse hook fails
>       debugPrint("Failed to set mouse hook: ${GetLastError()}");
>       _isMonitoring = false;
>       return;
>     }
> 
>     debugPrint("Monitoring started.");
>   }
> 
>   static void stopMonitoring() {
>     if (!_isMonitoring) return;
>     _isMonitoring = false;
> 
>     if (_keyboardHook != 0) {
>       UnhookWindowsHookEx(_keyboardHook);
>       _keyboardHook = 0; // Reset the hook
>     }
> 
>     if (_mouseHook != 0) {
>       UnhookWindowsHookEx(_mouseHook);
>       _mouseHook = 0; // Reset the hook
>     }
> 
>     debugPrint("Monitoring stopped.");
>   }
> 
>   static int _keyboardProc(int nCode, int wParam, int lParam) {
>     debugPrint("Keyboard event: wParam=$wParam, lParam=$lParam");
>     if (nCode >= 0) {
>       debugPrint("Keyboard event: wParam=$wParam, lParam=$lParam");
>       _keyBoardState.value = KeyboardState.active;
>     }
>     return CallNextHookEx(_keyboardHook, nCode, wParam, lParam);
>   }
> 
>   static int _mouseProc(int nCode, int wParam, int lParam) {
>     if (nCode >= 0) {
>       debugPrint("Mouse event: wParam=$wParam, lParam=$lParam");
>       _mouseState.value = MouseState.active;
>     }
>     return CallNextHookEx(_mouseHook, nCode, wParam, lParam);
>   }
> ```
> 
> I am not getting any event what will be the issue ?

@halildurmus  can you update this sample where I am doing something wrong?

**Comment by halildurmus** [comment]
`SetWindowsHookEx` requires [message loop](https://en.wikipedia.org/wiki/Message_loop_in_Microsoft_Windows) to work, which is why you're not receiving any events.

You need to create a Win32 window and run a message loop in another isolate, similar to the [usb_drive.dart](https://github.com/halildurmus/win32/blob/main/examples/usb_drive.dart) example and the example I gave in https://github.com/halildurmus/win32/discussions/912.
